src/crypto/public_key/rsa.js

```diff
@@ -18,254 +18,220 @@
-      key.d = toBigInteger(jwk.d);
-      key.p = toBigInteger(jwk.p);
-      key.q = toBigInteger(jwk.q);
-      key.u = key.p.modInverse(key.q);
+      let jwk = webCrypto.exportKey('jwk', keyPair.privateKey);
+      jwk = await promisifyIE11Op(jwk, 'Error exporting RSA key pair.');
 
-      function toBigInteger(base64url) {
-        const base64 = base64url.replace(/\-/g, '+').replace(/_/g, '/');
-        const hex = util.hexstrdump(atob(base64));
-        return new BigInteger(hex, 16);
+      // parse raw ArrayBuffer bytes to jwk/json (WebKit/Safari/IE11 quirk)
+      if (jwk instanceof ArrayBuffer) {
+        jwk = JSON.parse(String.fromCharCode.apply(null, new Uint8Array(jwk)));
       }
 
+      // map JWK parameters to BN
+      key = {};
+      key.n = b64toBN(jwk.n);
+      key.e = E;
+      key.d = b64toBN(jwk.d);
+      key.p = b64toBN(jwk.p);
+      key.q = b64toBN(jwk.q);
+      key.u = key.p.invm(key.q);
       return key;
     }
 
-    //
-    // JS code
-    //
-
-    return new Promise(function(resolve) {
-      const key = new KeyObject();
-      const rng = new SecureRandom();
-      const qs = B >> 1;
-      key.e = parseInt(E, 16);
-      key.ee = new BigInteger(E, 16);
+    // RSA keygen fallback using 40 iterations of the Miller-Rabin test
+    // See https://stackoverflow.com/a/6330138 for justification
+    let p = prime.randomProbablePrime(B - (B >> 1), E, 40);
+    let q = prime.randomProbablePrime(B >> 1, E, 40);
 
-      for (;;) {
-        for (;;) {
-          key.p = new BigInteger(B - qs, 1, rng);
-          if (key.p.subtract(BigInteger.ONE).gcd(key.ee).compareTo(BigInteger.ONE) === 0 && key.p.isProbablePrime(10)) {
-            break;
-          }
-        }
-        for (;;) {
-          key.q = new BigInteger(qs, 1, rng);
-          if (key.q.subtract(BigInteger.ONE).gcd(key.ee).compareTo(BigInteger.ONE) === 0 && key.q.isProbablePrime(10)) {
-            break;
-          }
-        }
-        if (key.p.compareTo(key.q) <= 0) {
-          const t = key.p;
-          key.p = key.q;
-          key.q = t;
-        }
-        const p1 = key.p.subtract(BigInteger.ONE);
-        const q1 = key.q.subtract(BigInteger.ONE);
-        const phi = p1.multiply(q1);
-        if (phi.gcd(key.ee).compareTo(BigInteger.ONE) === 0) {
-          key.n = key.p.multiply(key.q);
-          key.d = key.ee.modInverse(phi);
-          key.dmp1 = key.d.mod(p1);
-          key.dmq1 = key.d.mod(q1);
-          key.u = key.p.modInverse(key.q);
-          break;
-        }
-      }
-
-      resolve(key);
-    });
-  }
+    if (p.cmp(q) < 0) {
+      [p, q] = [q, p];
+    }
 
-  this.encrypt = encrypt;
-  this.decrypt = decrypt;
-  this.verify = verify;
-  this.sign = sign;
-  this.generate = generate;
-  this.keyObject = KeyObject;
-}
+    const phi = p.subn(1).mul(q.subn(1));
+    return {
+      n: p.mul(q),
+      e: E,
+      d: E.invm(phi),
+      q: q,
+      p: p,
+      // dq: d.mod(q.subn(1)),
```

Comment:
What was/is the purpose of these? We should probably either return these values for both the fallback and webcrypto or neither. But they don't seem necessary to me.

Reply:
Do you mean dq and dp? They're for faster RSA encryption and verification, but OpenPGP doesn't support them. I'll leave it there for the sake of reference.